### PR TITLE
Add a new feature for exclusive options and required at_least_one option

### PR DESCRIPTION
**Pull request by nstoym**
🌈
I've added a new feature for exclusive options and required at least one option.
For exclusive options,

``` ruby
exclusive do
  option :one
  option :two
end
```

You cann't give option `--one` and `--two` at the same time.
If you give wrong options, it shows an error as follows:

```
Found exclusive options '--one`, `--two`
```

For required at least one option:

``` ruby
at_least_one do
  option :three
  option :four
end
```

You must give  at least one option `--three` or `--four`.
This error message:

```
Not found at least one of required options `--three`, `--four`
```

`#class_exclusive` and `#class_at_least_one` methods can be used for `class_option`.
If `#excusive`/`#class_exclusive` or `#at_least_one`/`#class_at_least_one` are defined, the help command show you the relation of options after the`Options:` list as follows:

```
Exclusive Options:
  --one   --two

Required At Least One:
  --three  --four
```

### CODE REVIEW

**Comment by nstoym** [comment]
I can not buid test enviroments for Ruby 1.8.7 by `bundle install` for the following install error.

```
Gem::InstallError: celluloid requires Ruby version >= 1.9.2.
An error occurred while installing celluloid (0.16.0), and Bundler cannot continue.
Make sure that `gem install celluloid -v '0.16.0'` succeeds before bundling.
```
So, I need to insert the dependency in Gemfile.

```
platforms :ruby_18 do
  gem 'celluloid', '0.11.0'
end
```
I don't know whether this is correct. If you continue to support Ruby 1.8.7, I think it is necessary to fix this issue.

**Comment by sferik** [comment]
This looks good but it contains merge conflicts. Would you mind pulling from the upstream `master` and rebasing your branch?

**Comment by nstoym** [comment]
Hi sferik!

Thanks for your reply.
I will rebase soon.

> Can you please look into why this doesn’t produce an error message on Ruby 1.8.7?

This error message happened when I built test environment of Ruby 1.8.7 by using command `bundle install`.
"guard-rspec" gem in Gemfile depends on "celluloid" indirectly.
http://gemgraph.herokuapp.com/page?gem_name=guard-rspec&version=

"celluloid" gem requires Ruby >= 1.9.2 from version 0.12.0.
https://github.com/celluloid/celluloid/blob/v0.12.0/celluloid.gemspec

If we need "guard-rspec" gem, we should fix the Gemfile for new developers who don't have environment of Ruby 1.8.7.

**Comment by sferik** [comment]
> If we need "guard-rspec" gem, we should fix the Gemfile for new developers who don't have environment of Ruby 1.8.7.

We definitely don’t need `guard-rspec` in the `Gemfile`. I’ll remove it.

File: lib/thor.rb [hunk 1]
```diff
@@ -160,6 +160,97 @@ def disable_class_options
       @disable_class_options = true
     end
 
+    # Returns this class exclusive options array set.
+    #
+    # ==== Rturns
+    # Array[Array[Thor::Option.name]]
+    #
+    def method_exclusive_option_names
```
Comment (rafaelfranca) [inline lib/thor.rb]: This method should be protected and nodoc

File: lib/thor.rb [hunk 2]
```diff
@@ -160,6 +160,97 @@ def disable_class_options
...
+      @method_exclusive_option_names ||=[]
```
Comment (rafaelfranca) [inline lib/thor.rb]: Space after the =

File: lib/thor.rb [hunk 3]
```diff
@@ -160,6 +160,97 @@ def disable_class_options
...
+    end
+
+    # Adds and declareds option group for exclusive options in the
```
Comment (rafaelfranca) [inline lib/thor.rb]: `declares`

File: lib/thor.rb [hunk 4]
```diff
@@ -160,6 +160,97 @@ def disable_class_options
...
+    # block and arguments. You can declare options as the outside of the block.
+    # If :for is given as option,
+    # it allows you to change the options from a prvious defined command.
```
Comment (rafaelfranca) [inline lib/thor.rb]: `previous`

File: lib/thor.rb [hunk 5]
```diff
@@ -160,6 +160,97 @@ def disable_class_options
...
+    #
+    # ==== Parameters
+    # Array[Thor::Option.name]
+    # options<Hash>:: :for is applied for previous defined command.
+    #
+    # ==== Examples
+    #
+    #   exclusive do
```
Comment (rafaelfranca) [inline lib/thor.rb]: Could you show an example with the for option?

File: lib/thor.rb [hunk 6]
```diff
@@ -160,6 +160,97 @@ def disable_class_options
...
+    #     option :two
+    #    end
+    #
+    # Or
+    #
+    #   option :one
+    #   option :two
+    #   exclusive :one, :two
+    #
+    # If you give "--one" and "--two" at the same time.
```
Comment (rafaelfranca) [inline lib/thor.rb]: No need the `.` in the end.

File: lib/thor.rb [hunk 7]
```diff
@@ -160,6 +160,97 @@ def disable_class_options
...
+                                    :method_exclusive_option_names, *args, &block)
+    end
+    alias_method :exclusive, :method_exclusive
+
+    # Returns this class at least one of required options array set.
+    #
+    # ==== Rturns
+    # Array[Array[Thor::Option.name]]
+    #
+    def method_at_least_one_option_names
```
Comment (rafaelfranca) [inline lib/thor.rb]: This method should be protected and nodoc

File: lib/thor.rb [hunk 8]
```diff
@@ -160,6 +160,97 @@ def disable_class_options
...
+      @method_at_least_one_option_names ||=[]
+    end
+
+    # Adds and declareds option group for required at least one of options in the
```
Comment (rafaelfranca) [inline lib/thor.rb]: declares

File: lib/thor.rb [hunk 9]
```diff
@@ -160,6 +160,97 @@ def disable_class_options
...
+    #     option :two
+    #    end
+    #
+    # Or
+    #
+    #   option :one
+    #   option :two
+    #   at_least_one :one, :two
+    #
+    # If you do not give "--one" and "--two".
```
Comment (rafaelfranca) [inline lib/thor.rb]: No need the `.` in the end here

File: lib/thor.rb [hunk 10]
```diff
@@ -219,6 +315,26 @@ def printable_commands(all = true, subcommand = false)
     end
     alias_method :printable_tasks, :printable_commands
 
+    def print_exclusive_options(shell, command = nil)
+      opts = []
+      opts  = command.method_exclusive_option_names unless command.nil?
+      opts += class_exclusive_option_names
+      unless opts.empty?
+        shell.say "Exclusive Options:"
+        shell.print_table(opts.map{ |ex| ex.map{ |e| "--#{e}"}}, :indent => 2 )
```
Comment (rafaelfranca) [inline lib/thor.rb]: no need the space before the `)`

File: lib/thor.rb [hunk 11]
```diff
@@ -219,6 +315,26 @@ def printable_commands(all = true, subcommand = false)
...
+        shell.say
+      end
+    end
+    def print_at_least_one_required_options(shell, command = nil)
+      opts = []
+      opts = command.method_at_least_one_option_names unless command.nil?
+      opts += class_at_least_one_option_names
+      unless opts.empty?
+        shell.say "Required At Least One:"
+        shell.print_table(opts.map{ |ex| ex.map{ |e| "--#{e}"}}, :indent => 2 )
```
Comment (rafaelfranca) [inline lib/thor.rb]: No need the space before the `)`

File: lib/thor/base.rb [hunk 12]
```diff
@@ -62,8 +62,20 @@ def initialize(args = [], local_options = {}, config = {}) # rubocop:disable Met
...
       # a list of arguments that weren't declared.
-      stop_on_unknown = self.class.stop_on_unknown_option? config[:current_command]
-      opts = Thor::Options.new(parse_options, hash_options, stop_on_unknown)
+      current_command = config[:current_command]
+      stop_on_unknown = self.class.stop_on_unknown_option? current_command
+
+      # Give a relation of options.
+      # After parsing, Thor::Options check whether right relations are kept
+      relations = {:exclusive_option_names => [], :at_least_one_option_names => []}
+      unless current_command.nil?
```
Comment (rafaelfranca) [inline lib/thor/base.rb]: I think here a if/else would be clear

File: lib/thor/base.rb [hunk 13]
```diff
@@ -239,6 +251,24 @@ def arguments
         @arguments ||= from_superclass(:arguments, [])
       end
 
+      # Returns this class exclusive options array set, looking up in the ancestors chain.
+      #
+      # ==== Rturns
+      # Array[Array[Thor::Option.name]]
+      #
+      def class_exclusive_option_names
```
Comment (rafaelfranca) [inline lib/thor/base.rb]: Those methods should be protected and nodoc

File: lib/thor/base.rb [hunk 14]
```diff
@@ -274,6 +304,67 @@ def class_option(name, options = {})
         build_option(name, options, class_options)
       end
 
+      # Adds and declareds option group for exclusive options in the
```
Comment (rafaelfranca) [inline lib/thor/base.rb]: declares

File: lib/thor/base.rb [hunk 15]
```diff
@@ -274,6 +304,67 @@ def class_option(name, options = {})
...
+      #
+      # If you give "--one" and "--two" at the same time.
+      # ExclusiveArgumentsError will be raised.
+      #
+      def class_exclusive(*args, &block)
+        register_options_relation_for(:class_options,
+                                      :class_exclusive_option_names, *args, &block)
+      end
+
+      # Adds and declareds option group for required at least one of options in the
```
Comment (rafaelfranca) [inline lib/thor/base.rb]: declares

File: lib/thor/base.rb [hunk 16]
```diff
@@ -651,6 +742,34 @@ def initialize_added #:nodoc:
       def dispatch(command, given_args, given_opts, config) #:nodoc:
         fail NotImplementedError
       end
+
+      # Register a relation of options for target(method_option/class_option)
+      # by args and block.
+      def register_options_relation_for( target, relation, *args, &block)
```
Comment (rafaelfranca) [inline lib/thor/base.rb]: No need the space after the `(`

File: lib/thor/base.rb [hunk 17]
```diff
@@ -651,6 +742,34 @@ def initialize_added #:nodoc:
...
+      # of before and after by block evaluation.
+      def built_option_names(target, opt = {}, &block)
+        before = command_scope_member(target, opt).map{ |k,v| v.name }
+        instance_eval(&block)
+        after  = command_scope_member(target, opt).map{ |k,v| v.name }
+        after - before
+      end
+
+      # Get command scope member by name.
+      def command_scope_member( name, options = {} ) #:nodoc:
```
Comment (rafaelfranca) [inline lib/thor/base.rb]: no need the spaces around the ()

File: lib/thor/base.rb [hunk 18]
```diff
@@ -651,6 +742,34 @@ def initialize_added #:nodoc:
...
+        if options[:for]
+          find_and_refresh_command(options[:for]).send(name)
+        else
+          send( name )
```
Comment (rafaelfranca) [inline lib/thor/base.rb]: same here about the space

File: lib/thor/command.rb [hunk 19]
```diff
@@ -62,6 +63,12 @@ def formatted_usage(klass, namespace = true, subcommand = false)
       # Strip and go!
       formatted.strip
     end
+    def method_exclusive_option_names
```
Comment (rafaelfranca) [inline lib/thor/command.rb]: Missing a blank line above this method and below it

**Comment by IotaSpencer** [comment]
how similar is the code base if i were to update this to the latest thor codebase, as this is one of the features on my wish list for ruby command line frameworks

**Comment by rafaelfranca** [comment]
Merged in [291d93d](https://github.com/rails/thor/commit/291d93d6cb9f81123453f86292974bb7f0621300). Thanks!

**Pull request closed without merging**